File: docs/src/main/scala/sexamples/simulation/basic/BasicSimulation.scala
package sexamples.simulation.basic

import se.sics.kompics.sl._
import se.sics.kompics.sl.simulator._
import se.sics.kompics.simulator.{SimulationScenario => JSimulationScenario}
import scala.concurrent.duration._

case class SimpleEvent(num: Long = 0L) extends KompicsEvent;

object BasicSimulation {
  import Distributions._
  // needed for the distributions, but needs to be initialised after setting the seed
  implicit val random = JSimulationScenario.getRandom();

  // #gen1
  val simpleEventGen1 = Op { (_: Unit) =>
    SimpleEvent()
  };
  // #gen1

  // #gen2
  val simpleEventGen2 = Op { (param: java.lang.Long) =>
    SimpleEvent(param)
  };
  // #gen2

  val _ignoreThisIsJustForDocs = {
    // #gen1
    /* ... */
    raise(1000, simpleEventGen1).arrival(constant(2.seconds))
    // #gen1
    // #gen2
    /* ... */
    raise(1000, simpleEventGen2, uniform(1000L, 2000L)).arrival(constant(2.seconds))
    // #gen2
    ()
  }

  // #scenario
  val scenario = raise(1000, simpleEventGen1)
    .arrival(constant(2.seconds))
    .andThen(1.second)
    .afterTermination(
      raise(1000, simpleEventGen2, uniform(1000L, 2000L))
        .arrival(constant(2.seconds))
    )
    .andThen(2.seconds)
    .afterTermination(Terminate);
  // #scenario
}

File: docs/src/main/java/jexamples/simulation/basic/BasicSimulation.java
package jexamples.simulation.basic;

import se.sics.kompics.simulator.adaptor.Operation;
import se.sics.kompics.simulator.adaptor.Operation1;
import se.sics.kompics.simulator.SimulationScenario;

public class BasicSimulation {

  // #gen1
  static Operation<SimpleEvent> simpleEventGen1 =
      new Operation<SimpleEvent>() {

        @Override
        public SimpleEvent generate() {
          return new SimpleEvent();
        }
      };
  // #gen1

  // #gen2
  static Operation1<SimpleEvent, Long> simpleEventGen2 =
      new Operation1<SimpleEvent, Long>() {

        @Override
        public SimpleEvent generate(Long param) {
          return new SimpleEvent(param);
        }
      };
  // #gen2

  public static SimulationScenario scenario() {
    SimulationScenario scen =
        new SimulationScenario() {
          {
            // #gen1
            StochasticProcess p1 =
                new StochasticProcess() {
                  {
                    eventInterarrivalTime(constant(2000));
                    raise(1000, simpleEventGen1);
                  }
                };
            // #gen1

            // #gen2
            StochasticProcess p2 =
                new StochasticProcess() {
                  {
                    eventInterarrivalTime(constant(2000));
                    raise(1000, simpleEventGen2, uniform(1000L, 2000L));
                  }
                };
            // #gen2

            // #scenario
            p1.start();
            p2.startAfterTerminationOf(1000, p1);
            terminateAfterTerminationOf(1000, p2);
            // #scenario
          }
        };
    return scen;
  }
}
